scripts/copy-tables.py

```diff
@@ -0,0 +1,207 @@
+
+    assert (
+        to_local_table in all_tables
+    ), f"{to_local_table} needs to be created before {table}"
+    assert (
+        from_local_table in all_tables
+    ), f"{from_local_table} needs to be created before {table}"
+    print("...local tables found, mv check complete !\n")
+
+
+def copy_tables(
+    source_client: Client,
+    target_client: Client,
+    database: str,
+    execute: bool,
+    tables: Optional[Sequence[str]],
+) -> None:
+    """
+    When adding a replica to a clickhouse cluster, that node will not have any tables
+    created yet. In order to get the up to date table statements, we need to copy the
+    create table statements and run them on the new node(s).
+
+    The source_client is the clickhouse node that you are copying the table statements
+    from. The target_client is the clickhouse node is the node you want to create the
+    tables on.
+
+    The table engines that have some extra verification are materialized views, and
+    any of the replicated tables that are listed in the clickhouse docs:
+    https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replication/
+
+    By default running the script will only print out the CREATE TABLE statement, to
+    actually have the CREATE TABLE statement run, use --execute.
+
+    Consider doing one table at a time if you have tables that will take a while to
+    replicate and might increase the cluster's normalized load considerably.
+    """
+
+    if not tables:
+        tables = [result[0] for result in source_client.execute("SHOW TABLES")]
+
+    show_table_statments = OrderedDict()
+
+    for name in tables:
+        ((engine,),) = source_client.execute(
+            f"SELECT engine FROM system.tables WHERE name = '{name}'"
+        )
+
+        ((curr_create_table_statement,),) = source_client.execute(
+            f"SHOW CREATE TABLE {database}.{name}"
+        )
+
+        if engine == "MaterializedView":
+            print("\nMaterialized View Check:")
+            verify_local_tables_exist_from_mv(
+                target_client, curr_create_table_statement, name
+            )
+
+        if engine.startswith("Replicated"):
+            print("\nReplicated Table Check:")
+            verify_zk_replica_path(source_client, curr_create_table_statement, name)
+
+        show_table_statments[name] = curr_create_table_statement
+
+    ((_, target_replica), (_, target_shard)) = target_client.execute(
+        "SELECT macro, substitution FROM system.macros"
+    )
+    for table_name, statement in show_table_statments.items():
+        print(
+            f"creating {table_name}... on replica: {target_replica}, shard: {target_shard}"
+        )
+        if execute:
+            target_client.execute(statement)
+            print(f"created {table_name} !")
+            return
+        print(f"\ncreate table statement: \n {statement}\n")
+
+
+if __name__ == "__main__":
```

Comment:
Is there a way to test this functionality during CI? I would recommend spending sometime figuring out how this can be tested. An example could be bring up a cluster with 2 nodes and create some tables (raw and materialized) on one of the nodes. Then run the binary to copy those tables on the second node and verify that the tables exist.

Reply:
We could do something similar to test_distributed_migrations/test_runner_add_node.py where we add clickhouse--04 to the cluster. Also would be good to have unit tests for verify_zk_path functions to check the regex matching.